pkg/storage/storage.go

```diff
@@ -767,18 +788,46 @@ func (is *ImageStore) FullBlobUpload(repo string, body io.Reader, digest string)
 		return "", -1, errors.ErrBadBlobDigest
 	}
 
-	dir := path.Join(is.rootDir, repo)
-	dir = path.Join(dir, "blobs")
-	dir = path.Join(dir, dstDigest.Algorithm().String())
+	dir := path.Join(is.rootDir, repo, "blobs", dstDigest.Algorithm().String())
 	_ = os.MkdirAll(dir, 0755)
 	dst := is.BlobPath(repo, dstDigest)
 
-	// move the blob from uploads to final dest
-	_ = os.Rename(src, dst)
+	is.dedupeBlob(src, dstDigest, dst)
 
 	return uuid, n, err
 }
 
+func (is *ImageStore) dedupeBlob(src string, dstDigest godigest.Digest, dst string) {
+	dstRecord, err := is.cache.GetBlob(dstDigest.String())
+	if err != nil && err != cache.ErrCacheMiss {
+		is.log.Panic().Err(err).Str("blobPath", dst).Msg("unable to lookup blob record")
+	}
+
+	if dstRecord == "" {
+		if err := is.cache.PutBlob(dstDigest.String(), dst); err != nil {
+			is.log.Panic().Err(err).Str("blobPath", dst).Msg("unable to insert blob record")
+		}
+
+		// move the blob from uploads to final dest
+		_ = os.Rename(src, dst)
```

Comment:
If we risk actually losing blobs here, logging the error seems worthwhile.

Reply:
Handling the return error here and one other place.